test/canary/canary_test.go

```diff
@@ -32,46 +30,83 @@ func init() {
 	environment.RegisterEnvironmentMetaDataFlags(envMetaDataStrings)
 }
 
-// This copies the canary config, starts the agent, and confirm version is correct.
-// Sanity test runs earlier in the terraform code execution
+// TestCanary verifies customers can download the agent, install it, and it "works".
+// Reports metrics on the number of download failures, install failures, and start failures.
 func TestCanary(t *testing.T) {
-	// Canary set up
-	downloadVersionFile()
+	installerFilePath := "./downloaded_cwa.rpm"
+	err := downloadInstaller(installerFilePath)
+	reportMetric(t, "DownloadFail", err)
+
+	// Don't care if uninstall fails. Agent might not be installed anyways.
+	_ = common.UninstallAgent()
+
+	err = common.InstallAgent(installerFilePath)
+	reportMetric(t, "InstallFail", err)
+
 	common.CopyFile(configInputPath, configOutputPath)
-	common.StartAgent(configOutputPath, true)
+	err = common.StartAgent(configOutputPath, false)
+	reportMetric(t, "StartFail", err)
 
-	// Version validation
-	expectedVersion, err := os.ReadFile(downloadAgentVersionPath)
-	if err != nil {
-		t.Fatalf("Failure reading downloaded version file %s err %v", downloadAgentVersionPath, err)
-	}
-	actualVersion, err := os.ReadFile(installAgentVersionPath)
-	if err != nil {
-		t.Fatalf("Failure reading installed version file %s err %v", installAgentVersionPath, err)
+	actualVersion, _ := os.ReadFile(installAgentVersionPath)
+	expectedVersion, _ := getVersionFromS3()
+	if expectedVersion != string(actualVersion) {
+		err = errors.New("agent version mismatch")
 	}
-	assert.Equal(t, string(expectedVersion), string(actualVersion))
+	reportMetric(t, "VersionFail", err)
```

Comment:
It good we want to be robust but IMO, test should be generalize as much as possible (if these states failed, it would show the error in the github action console, then why we should look at these metrics?)